File: lib/plotting_r/stages/_shared.R
# Shared helpers for R stage modules (Stage 13/14 parity with the
# Python-side `lib/plotting/stages/_shared.py`).
#
# Stage modules read manifest entries to find a TSV path relative to
# `outputs_dir`, then load specific columns by header-name aliases.
# Centralising the boilerplate keeps each stage module under ~100 LOC
# and matches the Python contract: missing tables / unparseable
# headers raise the same family of errors so `core.R::swfc_generate`
# records them per-figure.

if (!exists("swfc_register_figure")) {
  stop("source runtime/plotting_r/core.R before _shared.R")
}

# Resolve a manifest entry whose value is a relative path against
# `outputs_dir`. Returns NULL when any segment is missing or the
# resolved file does not exist on disk.
.swfc_manifest_path <- function(manifest, outputs_dir, ...) {
  keys <- c(...)
  cur <- manifest
  for (k in keys) {
    if (!is.list(cur) || is.null(cur[[k]])) return(NULL)
    cur <- cur[[k]]
  }
  if (!is.character(cur) || length(cur) != 1L) return(NULL)
  p <- normalizePath(file.path(outputs_dir, cur), mustWork = FALSE)
  if (!file.exists(p)) return(NULL)
  p
}

# Read a TSV (gzip-aware via R's connection inference) and return the
# requested columns as a data.frame. `columns` is a named list mapping
# the canonical column name to a vector of accepted header aliases.
# Canonical names ending in `?` mark optional columns — when no alias
# matches, the returned frame omits the column rather than failing.
.swfc_load_tsv <- function(path, columns) {
  df <- tryCatch(
    utils::read.delim(path, stringsAsFactors = FALSE, check.names = FALSE,
                      comment.char = ""),
    error = function(e) NULL
  )
  if (is.null(df) || nrow(df) == 0L) return(NULL)
  header_lc <- tolower(colnames(df))
  out <- list()
  for (canonical in names(columns)) {
    aliases <- tolower(columns[[canonical]])
    idx <- which(header_lc %in% aliases)
    if (length(idx) == 0L) {
      if (endsWith(canonical, "?")) next
      return(NULL)
    }
    name <- sub("\\?$", "", canonical)
    out[[name]] <- df[[idx[1]]]
  }
  as.data.frame(out, stringsAsFactors = FALSE)
}

File: crates/ecaa-conformance/tests/fixtures/minimal-package-real/runtime/plotting/stages/_shared.py
"""Shared helpers for stage modules — TSV/JSON manifest readers.

Stage 12-14 introduces ~16 new stage modules whose data-loading logic
is identical: read a path from the manifest, open the TSV (gzip-aware),
return columns as numpy arrays. Centralising the boilerplate keeps each
stage module thin (<100 LOC) and tightens the contract that bad inputs
fall through to a typed FileNotFoundError that `core.generate()`
records on the per-figure `errors` map.
"""

from __future__ import annotations

import gzip
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

import numpy as np


def open_text(path: Path):
    """Open a TSV path with gzip awareness."""
    if str(path).endswith(".gz"):
        return gzip.open(path, "rt")
    return open(path, "rt")


def find_col(header: List[str], candidates: Tuple[str, ...]) -> Optional[int]:
    lower = [h.lower() for h in header]
    for cand in candidates:
        try:
            return lower.index(cand.lower())
        except ValueError:
            continue
    return None


def manifest_path(manifest: Dict[str, Any], outputs_dir: Path, *keys: str) -> Optional[Path]:
    """Walk `manifest[keys[0]][keys[1]]...` and resolve the leaf string
    against `outputs_dir`. Returns None if any segment is missing or
    not a string. Used to keep stage-module logic free of nested
    .get(...).get(...) dance.
    """
    cur: Any = manifest
    for k in keys:
        if not isinstance(cur, dict):
            return None
        cur = cur.get(k)
    if not isinstance(cur, str):
        return None
    p = (outputs_dir / cur).resolve()
    return p if p.exists() else None


def load_tsv_columns(
    path: Path, columns: Dict[str, Tuple[str, ...]]
) -> Optional[Dict[str, np.ndarray]]:
    """Load a TSV file, returning the requested columns by candidate
    header names. `columns` maps the canonical name (used by the
    primitive) to a tuple of acceptable header aliases.

    Returns None if the file can't be opened or any required column is
    missing. Required columns are those whose canonical name does not
    end in `?` (which marks an optional column).

    Example:
        load_tsv_columns(path, {
            "chrom": ("chrom", "chromosome", "#CHROM"),
            "pos": ("pos", "position", "POS"),
            "pvalue": ("pvalue", "P", "p_value"),
            "gene?": ("gene", "Gene", "symbol"),
        })
    """
    try:
        with open_text(path) as f:
            header = f.readline().rstrip("\n").split("\t")
            indices: Dict[str, Optional[int]] = {}
            for canonical, aliases in columns.items():
                indices[canonical] = find_col(header, aliases)
            for canonical, idx in indices.items():
                if not canonical.endswith("?") and idx is None:
                    return None
            data: Dict[str, List[str]] = {k: [] for k in columns}
            for line in f:
                parts = line.rstrip("\n").split("\t")
                for canonical, idx in indices.items():
                    if idx is None:
                        continue
                    val = parts[idx] if idx < len(parts) else ""
                    data[canonical].append(val)
        out: Dict[str, np.ndarray] = {}
        for canonical, vals in data.items():
            stripped = canonical.rstrip("?")
            if not vals:
                # Optional missing column — leave out.
                continue
            try:
                out[stripped] = np.asarray(vals, dtype=float)
            except ValueError:
                out[stripped] = np.asarray(vals, dtype=object)
        return out
    except OSError:
        return None
